Unused imports

```
pylint *.py | grep -i unused

api.py:5:0: W0611: Unused import sys (unused-import)
api.py:7:0: W0611: Unused Counter imported from collections (unused-import)
ui.py:6:0: W0611: Unused import sys (unused-import)
ui.py:7:0: W0611: Unused import yaml (unused-import)
ui.py:13:0: W0611: Unused numpy imported as np (unused-import)
```
In general, I suggest to use `pylint` to check for things like that.

BTW `numpy` isn't in the packages installed at `Dockerfile`. Its installed by some other dependency?

Tip: When you import something directly in your code, you should add it in the dependencies you install.
Maybe tomorrow the package that installs this dependency changes and it won't be installed any more...

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -7,10 +7,8 @@ on:
 jobs:
   lint:
     runs-on: ubuntu-latest
-    name: Lint
+    name: Lint in Docker
     steps:
       - uses: actions/checkout@v3
-      - name: Lint Python
-        run: |
-          pip install --user pylint
-          pylint --max-line-length=120 --good-names="c,ct,e,ep,id,q,r" --disable="C0114,C0115,C0116" --extension-pkg-whitelist="pydantic" *.py
+      - name: Build Docker Image
+        run: ./Dockerfile

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -1,0 +1,16 @@
+name: Lint Code
+
+on:
+  - push
+  - pull_request
+
+jobs:
+  lint:
+    runs-on: ubuntu-latest
+    name: Lint
+    steps:
+      - uses: actions/checkout@v3
+      - name: Lint Python
+        run: |
+          pip install --user pylint
+          pylint --max-line-length=120 --good-names="c,ct,e,ep,id,q,r" --disable="C0114,C0115,C0116" --extension-pkg-whitelist="pydantic" *.py

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -6,6 +6,7 @@ on:
 
 jobs:
   lint:
+    if: github.event_name == 'push' || github.event.pull_request.head.repo.full_name != github.repository
     runs-on: ubuntu-latest
     name: Lint in Docker
     steps:

--- a/Dockerfile
+++ b/Dockerfile
@@ -8,10 +8,16 @@ WORKDIR /app
 CMD     ["./api.py"]
 
 RUN     pip install --no-cache-dir \
+            altair \
             "elasticsearch>=7.0.0,<8.0.0" \
             fastapi \
+            matplotlib \
+            pandas \
+            pydantic \
+            requests \
             streamlit \
             "uvicorn[standard]" \
-            wordcloud
+            wordcloud \
+            yaml
 
 COPY . ./

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,11 +1,11 @@
 #!/usr/bin/env -S docker image build -t colsearch . -f
 
-FROM    python:3
+# Base image
+FROM    python:3 AS base
 
 ENV     STREAMLIT_BROWSER_GATHER_USAGE_STATS=false
 
 WORKDIR /app
-CMD     ["./api.py"]
 
 RUN     pip install --no-cache-dir \
             altair \
@@ -18,6 +18,21 @@ RUN     pip install --no-cache-dir \
             streamlit \
             "uvicorn[standard]" \
             wordcloud \
-            yaml
+            pyyaml
 
-COPY . ./
+COPY    . ./
+
+# Lint code
+FROM    base
+
+RUN     pip install --no-cache-dir pylint
+RUN     pylint *.py \
+            --max-line-length=120 \
+            --good-names="c,ct,e,ep,id,q,r" \
+            --disable="C0114,C0115,C0116" \
+            --extension-pkg-whitelist="pydantic"
+
+# Build image
+FROM    base
+
+CMD     ["./api.py"]

--- a/api.py
+++ b/api.py
@@ -1,18 +1,18 @@
 #!/usr/bin/env python3
 
+
 import base64
 import os
-import sys
 
-from collections import Counter
+from typing import Union
+from urllib.parse import quote_plus
+
 from elasticsearch import Elasticsearch
 from elasticsearch.exceptions import TransportError
 from fastapi import FastAPI, Request, Response, Body, HTTPException
 from fastapi.middleware.cors import CORSMiddleware
 from fastapi.responses import HTMLResponse, RedirectResponse
 from pydantic import BaseModel
-from typing import Union
-from urllib.parse import quote_plus
 
 from utils import load_config, env_to_list, list_to_enum
 
@@ -76,12 +76,12 @@ class PagedQuery(Query):
     resume: Union[str, None] = None
 
 
-def encode(str: str):
-    return base64.b64encode(str.encode(), b"-_").decode()
+def encode(strng: str):
+    return base64.b64encode(strng.encode(), b"-_").decode()
 
 
-def decode(str: str):
-    return base64.b64decode(str.encode(), b"-_").decode()
+def decode(strng: str):
+    return base64.b64decode(strng.encode(), b"-_").decode()
 
 
 def cs_basic_query(q: str):
@@ -230,14 +230,15 @@ def api_entrypoint(req: Request):
     """
     Link to the interactive API documentation
     """
+    href = f"{req.scope.get('root_path')}/{VERSION_1}/docs"
     return "\n".join(['<ul>',
-            f'<li><a href="{req.scope.get("root_path")}/{VERSION_1}/docs">Interactive API Docs ({VERSION_1})</a></li>',
-            '</ul>'])
+                      f'<li><a href="{href}">Interactive API Docs ({VERSION_1})</a></li>',
+                      '</ul>'])
 
 
 @app.get("/docs", response_class=RedirectResponse)
 @app.head("/docs", response_class=RedirectResponse)
-def api_entrypoint(req: Request):
+def api_entrypoint_docs(req: Request):
     """
     Redirect to recent API documentation
     """
@@ -246,7 +247,7 @@ def api_entrypoint(req: Request):
 
 @app.get("/redoc", response_class=RedirectResponse)
 @app.head("/redoc", response_class=RedirectResponse)
-def api_entrypoint(req: Request):
+def api_entrypoint_redoc(req: Request):
     """
     Redirect to recent API documentation
     """
@@ -259,7 +260,8 @@ def version_root(req: Request):
     """
     Links to various collections
     """
-    return "\n".join(['<ul>'] + [f'<li><a href="{req.scope.get("root_path")}/{col.value}">{col.value}</a></li>' for col in Collection] + ['</ul>'])
+    lis = [f'<li><a href="{req.scope.get("root_path")}/{col.value}">{col.value}</a></li>' for col in Collection]
+    return "\n".join(['<ul>'] + lis + ['</ul>'])
 
 
 @v1.get("/{collection}", response_class=HTMLResponse, tags=["info"])
@@ -269,10 +271,10 @@ def collection_root(collection: Collection, req: Request):
     Links to various collection API endpoints
     """
     return "\n".join(['<ul>',
-            f'<li><a href="{req.scope.get("root_path")}/{collection.value}/search">Search API</a></li>',
-            f'<li><a href="{req.scope.get("root_path")}/{collection.value}/terms">Related Terms API</a></li>',
-            f'<li><a href="{req.scope.get("root_path")}/{collection.value}/article">Artcile</a></li>',
-            '</ul>'])
+                      f'<li><a href="{req.scope.get("root_path")}/{collection.value}/search">Search API</a></li>',
+                      f'<li><a href="{req.scope.get("root_path")}/{collection.value}/terms">Related Terms API</a></li>',
+                      f'<li><a href="{req.scope.get("root_path")}/{collection.value}/article">Artcile</a></li>',
+                      '</ul>'])
 
 
 @v1.get("/{collection}/search", response_class=HTMLResponse, tags=["info"])
@@ -281,24 +283,28 @@ def search_root(collection: Collection, req: Request):
     """
     Links to various search API endpoints
     """
+    spath = f"{req.scope.get('root_path')}/{collection.value}/search"
     return "\n".join(['<ul>',
-            f'<li><a href="{req.scope.get("root_path")}/{collection.value}/search/overview">Search Overview</a></li>',
-            f'<li><a href="{req.scope.get("root_path")}/{collection.value}/search/result">Search Result</a></li>',
-            '</ul>'])
+                      f'<li><a href="{spath}/overview">Search Overview</a></li>',
+                      f'<li><a href="{spath}/result">Search Result</a></li>',
+                      '</ul>'])
 
 
 def _search_overview(collection: Collection, q: str, req: Request):
     res = ES.search(index=collection.name, body=cs_overview_query(q))
     if not res["hits"]["hits"]:
         raise HTTPException(status_code=404, detail="No results found!")
+    total = res["hits"]["total"]["value"]
+    tldsum = sum(item["doc_count"] for item in res["aggregations"]["tld"]["buckets"])
+    base = f'{req.base_url}{req.scope.get("root_path").lstrip("/")}'
     return {
         "query": q,
-        "total": max(res["hits"]["total"]["value"], sum(item["doc_count"] for item in res["aggregations"]["tld"]["buckets"])),
+        "total": max(total, tldsum),
         "topdomains": format_counts(res["aggregations"]["domain"]["buckets"]),
         "toptlds": format_counts(res["aggregations"]["tld"]["buckets"]),
         "toplangs": format_counts(res["aggregations"]["lang"]["buckets"]),
         "dailycounts": format_day_counts(res["aggregations"]["daily"]["buckets"]),
-        "matches": [format_match(h, f'{req.base_url}{req.scope.get("root_path").lstrip("/")}', collection.value) for h in res["hits"]["hits"]]
+        "matches": [format_match(h, base, collection.value) for h in res["hits"]["hits"]]
     }
 
 
@@ -323,16 +329,18 @@ def _search_result(collection: Collection, q: str, req: Request, resp: Response,
     res = ES.search(index=collection.name, body=cs_paged_query(q, resume))
     if not res["hits"]["hits"]:
         raise HTTPException(status_code=404, detail="No results found!")
+    base = f'{req.base_url}{req.scope.get("root_path").lstrip("/")}'
+    qurl = f"{base}/{collection.value}/search/result?q={quote_plus(q)}"
     if len(res["hits"]["hits"]) == config["maxpage"]:
         resume_key = encode(res["hits"]["hits"][-1]["sort"][0])
         resp.headers["x-resume-token"] = resume_key
-        resp.headers["link"] = f'<{req.base_url}{req.scope.get("root_path").lstrip("/")}/{collection.value}/search/result?q={quote_plus(q)}&resume={resume_key}>; rel="next"'
-    return [format_match(h, f'{req.base_url}{req.scope.get("root_path").lstrip("/")}', collection.value) for h in res["hits"]["hits"]]
+        resp.headers["link"] = f'<{qurl}&resume={resume_key}>; rel="next"'
+    return [format_match(h, base, collection.value) for h in res["hits"]["hits"]]
 
 
 @v1.get("/{collection}/search/result", tags=["data"])
 @v1.head("/{collection}/search/result", include_in_schema=False)
-def search_result_via_query_params(collection: Collection, q: str, req: Request, resp: Response, resume: Union[str, None] = None):
+def search_result_via_query_params(collection: Collection, q: str, req: Request, resp: Response, resume: Union[str, None] = None):  # pylint: disable=line-too-long # noqa: E501
     """
     Paged response of search result
     """
@@ -369,7 +377,9 @@ def term_field_root(collection: Collection, req: Request):
     """
     Links to various term fields
     """
-    return "\n".join(['<ul>'] + [f'<li><a href="{req.scope.get("root_path")}/{collection.value}/terms/{field.value}">{field.value}</a></li>' for field in TermField] + ['</ul>'])
+    tbase = f"{req.scope.get('root_path')}/{collection.value}/terms"
+    lis = [f'<li><a href="{tbase}/{field.value}">{field.value}</a></li>' for field in TermField]
+    return "\n".join(['<ul>'] + lis + ['</ul>'])
 
 
 @v1.get("/{collection}/terms/{field}", response_class=HTMLResponse, tags=["info"])
@@ -378,7 +388,9 @@ def term_aggr_root(collection: Collection, req: Request, field: TermField):
     """
     Links to various term aggregations
     """
-    return "\n".join(['<ul>'] + [f'<li><a href="{req.scope.get("root_path")}/{collection.value}/terms/{field.value}/{aggr.value}">{aggr.value}</a></li>' for aggr in TermAggr] + ['</ul>'])
+    fbase = f"{req.scope.get('root_path')}/{collection.value}/terms/{field.value}"
+    lis = [f'<li><a href="{fbase}/{aggr.value}">{aggr.value}</a></li>' for aggr in TermAggr]
+    return "\n".join(['<ul>'] + lis + ['</ul>'])
 
 
 @v1.get("/{collection}/terms/{field}/{aggr}", tags=["data"])
@@ -400,14 +412,14 @@ def get_terms_via_payload(collection: Collection, payload: Query, field: TermFie
 
 @v1.get("/{collection}/article/{id}", tags=["data"])
 @v1.head("/{collection}/article/{id}", include_in_schema=False)
-def get_article(collection: Collection, id: str, req: Request):
+def get_article(collection: Collection, id: str, req: Request):  # pylint: disable=redefined-builtin
     """
     Fetch an individual article record by ID
     """
     try:
         hit = ES.get(index=collection.name, id=id)
     except TransportError as e:
-        raise HTTPException(status_code=404, detail=f"An article with ID {id} not found!")
+        raise HTTPException(status_code=404, detail=f"An article with ID {id} not found!") from e
     return format_match(hit, f'{req.base_url}{req.scope.get("root_path").lstrip("/")}', collection.value, True)
 
 

--- a/ui.py
+++ b/ui.py
@@ -2,17 +2,16 @@
 
 import os
 import random
+
+from urllib.parse import quote_plus
+
 import requests
-import sys
-import yaml
 
 import matplotlib.pyplot as plt
 import altair as alt
 import streamlit as st
 import pandas as pd
-import numpy as np
 
-from urllib.parse import quote_plus
 from wordcloud import WordCloud
 
 from utils import load_config, env_to_list
@@ -30,11 +29,12 @@ st.set_page_config(page_title=config["title"], layout="wide")
 st.title(config["title"])
 
 
-#@st.cache(ttl=300)
-def load_data(c, q, ep="search/overview"):
-    r = requests.get(f"{config['apiurl']}/{c}/{ep}?q={quote_plus(q)}")
+# @st.cache(ttl=300)
+def load_data(cname, qstr, ep="search/overview"):
+    r = requests.get(f"{config['apiurl']}/{cname}/{ep}?q={quote_plus(qstr)}")
     if r.ok:
         return r.json()
+    return None
 
 
 qp = st.experimental_get_query_params()
@@ -74,7 +74,10 @@ cols[3].metric("Days", f"{len(ov['dailycounts']):,}")
 tbs = st.tabs(["Top Hits", "Data"])
 res = ["Title | Domain | Published | Archived | Language", ":---|:---|:---:|:---:|:---:"]
 for m in ov["matches"]:
-    res.append(f"[{m.get('title', 'UNKNOWN').replace('|', '&vert;')}]({m.get('archive_playback_url') or '#'}) | `{m.get('domain') or '~'}` | `{m.get('publication_date') or '~'}` | `{(m.get('capture_time') or '~')[:10]}` | `{m.get('language') or '~'}`")
+    t = m.get("title", "UNKNOWN").replace("|", "&vert;")
+    res.append(" | ".join([f"[{t}]({m.get('archive_playback_url') or '#'})",
+                           f"`{m.get('domain') or '~'}` | `{m.get('publication_date') or '~'}`",
+                           f"`{(m.get('capture_time') or '~')[:10]}` | `{m.get('language') or '~'}`"]))
 tbs[0].write("\n".join(res))
 tbs[1].write(ov["matches"])
 
@@ -114,7 +117,12 @@ for fld in ["title", "snippet"]:
             tbs = st.tabs([f"{aggr} {fld} terms".title(), "Data"])
             tt = load_data(col, q, f"terms/{fld}/{aggr}")
             if tt:
-                sample = (dict(random.sample(list(tt.items()), config["maxwc"])) if aggr == "rare" else dict(list(tt.items())[:config["maxwc"]])) if len(tt) > config["maxwc"] else tt
+                sample = tt
+                if len(tt) > config["maxwc"]:
+                    if aggr == "rare":
+                        sample = dict(random.sample(list(tt.items()), config["maxwc"]))
+                    else:
+                        sample = dict(list(tt.items())[:config["maxwc"]])
                 wc = WordCloud(background_color="white")
                 wc.generate_from_frequencies(sample)
                 fig, ax = plt.subplots()

--- a/utils.py
+++ b/utils.py
@@ -1,14 +1,15 @@
 import os
-import yaml
 
 from enum import Enum
+
+import yaml
 
 
 def load_config():
     conf = os.getenv("CONFIG", "config.yml")
     try:
-        return yaml.safe_load(open(conf))
-    except OSError as e:
+        return yaml.safe_load(open(conf, encoding="UTF-8"))
+    except OSError:
         return {}
 
 

